test/stream-monitor/stream_sources/amqp_source.py

```diff
@@ -108,47 +126,165 @@ def _match(self, other_event):
-        super(_AMQPQueueMonitor, self).__init__()
-        # let base class catch up with us statewise (we get created from
-        # inside the entire amqp-stream-monitor after begin). We do this here
-        # so that nothing can wander in as the queues are set up, etc.
+class _AMQPProcessor(StreamMonitorBaseClass):
+    def __init__(self, logs, tracker, start_at=None, transient=True):
+        self._logs = logs
+        super(_AMQPProcessor, self).__init__()
         self.handle_begin()
+        self.transient = transient
+        self.__tracker = tracker
+        self.__inbound_queue = gevent.queue.Queue()
+        self.__run_till = None
+        self.__tail_wait = None
+        self.__complete_on_success = False
+        tracker.add_processor(self, start_at=start_at)
+        self.__match_greenlet = gevent.spawn(self.__match_greenlet_run)
+        self.__match_greenlet.greenlet_name = 'processor-match-loop-gl'
+
+    def __match_greenlet_run(self):
+        self._logs.irl.debug('Starting to watch for my events %s', self)
+        results = StreamRunResults()
+
+        self._logs.irl.debug("cp0")
+        while self.__run_till is not None and self.__run_till > time.time():
+            tail_waiting = self.__tail_wait
+            try:
+                # timeout on peek call is needed to allow us to "notice" if our run-till
+                # or tail-time has been exceeded.
+                tracked = self.__inbound_queue.peek(timeout=0.1)
+                self._logs.idl.debug('%s peeked and got %s', self, tracked)
+            except gevent.queue.Empty:
+                tracked = None
+
+            if tracked is None:
+                if tail_waiting is not None and time.time() > tail_waiting:
+                    self._logs.irl.debug(' tail-wait timed out.')
+                    break
+                continue
+
+            res = self._match_groups.check_event(tracked)
+            if res is not None:
+                # actually remove item from queue now that we know we have a "hit"
+                self.__inbound_queue.get()
+                results.add_result(res)
+
+            if self.__complete_on_success:
+                res = self._match_groups.check_ending()
+                # currently, it's either an error (so we keep trying) or None
+                if res is None:
+                    results.add_result(res)
+                    self._logs.irl.debug('  processor is currently evaluating as a success point, so we are bailing')
+                    return results
+
+        self._logs.irl.debug('---exiting because of time---: %s -> %s', self, results)
+        res = self._match_groups.check_ending()
+        results.add_result(res)
+        self._logs.irl.debug('  final results from %s is %s', self, results)
+        return results
+
+    def start_finish(self, timeout, tail_timeout=1.0):
+        self._logs.irl.debug('start finish on %s called. timeout=%s, tail-timeout=%s', self, timeout, tail_timeout)
```

Comment:
It was actually more of a verb (the method name start_finish is a property built into log records). But if it jumps out, I'll change it.

If you are ok with the cp0 being removed in the next story, I'll put this fix off till then also: that way I can skip a round of rebases/re-tests.